packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/bedrock/create-model-customization-job.ts

```diff
@@ -0,0 +1,396 @@
+   */
+  readonly validationDataS3Uri: string[];
+  /**
+   * The IAM role that Amazon Bedrock can assume to perform tasks on your behalf.
+   *
+   * For example, during model training, Amazon Bedrock needs your permission to read input data from an S3 bucket,
+   * write model artifacts to an S3 bucket.
+   * To pass this role to Amazon Bedrock, the caller of this API must have the iam:PassRole permission.
+   *
+   * @default - use auto generated role
+   */
+  readonly role?: iam.IRole;
+  /**
+   * Configuration parameters for the private Virtual Private Cloud (VPC) that contains the resources you are using for this job.
+   *
+   * @default - no VPC configuration
+   */
+  readonly vpcConfig?: IBedrockCreateModelCustomizationJobVpcConfig;
+}
+
+/**
+ * A Step Functions Task to create model customization in Bedrock.
+ */
+export class BedrockCreateModelCustomizationJob extends sfn.TaskStateBase {
+
+  private static readonly SUPPORTED_INTEGRATION_PATTERNS: sfn.IntegrationPattern[] = [
+    sfn.IntegrationPattern.REQUEST_RESPONSE,
+    sfn.IntegrationPattern.RUN_JOB,
+  ];
+
+  protected readonly taskMetrics: sfn.TaskMetricsConfig | undefined;
+  protected readonly taskPolicies: iam.PolicyStatement[] | undefined;
+
+  private readonly integrationPattern: sfn.IntegrationPattern;
+  private _role: iam.IRole;
+
+  constructor(scope: Construct, id: string, private readonly props: BedrockCreateModelCustomizationJobProps) {
+    super(scope, id, props);
+    this.integrationPattern = props.integrationPattern ?? sfn.IntegrationPattern.REQUEST_RESPONSE;
+
+    this.validateStringLength('clientRequestToken', 1, 256, props.clientRequestToken);
+    this.validatePattern('clientRequestToken', /^[a-zA-Z0-9](-*[a-zA-Z0-9])*$/, props.clientRequestToken);
+    this.validateStringLength('customModelName', 1, 63, props.customModelName);
+    this.validatePattern('customModelName', /^([0-9a-zA-Z][_-]?)+$/, props.customModelName);
+    this.validateArrayLength('customModelTags', 0, 200, props.customModelTags);
+    this.validateStringLength('jobName', 1, 63, props.jobName);
+    this.validatePattern('jobName', /^[a-zA-Z0-9](-*[a-zA-Z0-9\+\-\.])*$/, props.jobName);
+    this.validateArrayLength('jobTags', 0, 200, props.jobTags);
+    this.validateArrayLength('validationDataS3Uri', 1, 10, props.validationDataS3Uri);
+    this.validateArrayLength('securityGroups', 1, 5, props.vpcConfig?.securityGroups);
+    this.validateArrayLength('subnets', 1, 16, props.vpcConfig?.subnets);
+
+    validatePatternSupported(this.integrationPattern, BedrockCreateModelCustomizationJob.SUPPORTED_INTEGRATION_PATTERNS);
+
+    this._role = this.renderBedrockCreateModelCustomizationJobRole();
+    this.taskPolicies = this.renderPolicyStatements();
+  }
+
+  /**
+   * The IAM role for the bedrock create model customization job
+   */
+  public get role(): iam.IRole {
+    return this._role;
+  }
+
+  private renderBedrockCreateModelCustomizationJobRole(): iam.IRole {
```

Comment:
Can you add a link to the related documentation for this role?

Reply:
I've added the link of the documentation touching about S3 access. 
I am referring to the display on the management console for the IAM policy regarding the VPC config settings and there may be no documentation about it.

<img width="847" alt="スクリーンショット 2024-04-01 18 06 18" src="https://github.com/aws/aws-cdk/assets/64848616/a154725b-0578-4ddb-a051-10ea16950e80">

Reply:
This [doc](https://docs.aws.amazon.com/bedrock/latest/userguide/vpc-model-customization.html) seems more related. Can you please update the documentation and the policy as necessary?

Reply:
Thanks.
I've updated the policy description.

Reply:
Following, point 1.
If no VPC or a VPC with public internet access is provided, this is correct.
Otherwise, if a VPC with no public internet access is provided, we should validate that the VPC has a gateway endpoint to S3 attached, and provide the bucket policy, VPC endpoint policy, and role permissions specified [here](https://docs.aws.amazon.com/bedrock/latest/userguide/vpc-model-customization.html).

Reply:
When dealing with subnets that cannot access the Internet (isolated subnet), I've come to think that it might be best not to handle the following within AWS CDK:
- Verification of endpoint existence
- Configuration of endpoint policies
- Configuration of bucket policies

Starting with the verification of endpoint existence, the [documentation](https://docs.aws.amazon.com/bedrock/latest/userguide/vpc-model-customization.html) mentions the use of S3 gateway endpoints, but I believe S3 interface endpoints might also be used in practice. Indeed, the documentation refers to private access via PrivateLink, i.e., interface endpoints, at the beginning. It seems more appropriate for users to take responsibility for ensuring access routes to S3 from isolated subnets, including setting up appropriate security groups as arguments.

As for endpoint policies, I view these as optional settings to enhance security rather than mandatory configurations. The [documentation](https://docs.aws.amazon.com/bedrock/latest/userguide/vpc-model-customization.html#train-vpc-policy) explicitly permits certain job-related permissions, but such configurations implicitly deny all other S3 operations. Implicitly adding this policy through CDK might affect other uses of the same VPC endpoint.

Similarly, for bucket policies, there could be implications, such as denying operations from outside the VPC for buckets intended for output data, which might be accessed externally.

I apologize for the lengthy discussion but would greatly appreciate any thoughts on this matter. Thank you for your continued support.

Reply:
> the documentation mentions the use of S3 gateway endpoints, but I believe S3 interface endpoints might also be used in practice.

Agree, thanks for making the point.

> I think it's valid to verify the existence of either an Interface or Gateway type endpoint

Yes, only if the VPC cannot access the internet.

> Implicitly adding this policy through CDK might affect other uses of the same VPC endpoint.

🤔 
I think we can add the bucket policy restricting access to the VPC without issues.
For the VPC endpoint policy, what about adding a condition that enforces it only on the task role? ([docs](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_condition-keys.html#condition-keys-principalarn))

Reply:
> Yes, only if the VPC cannot access the internet.
Sure. I'll add verification when all `vpcConfig.subnets` are isolated.

How can I retrieve the endpoint resource on this construct?
My only solution is to use `App` or `Stack.node.findAll()` because `IVpc` does not have `endpoints` properties.

```ts
    if (this.props.vpcConfig) {
      if (this.props.vpcConfig.subnets) {
        this.subnets = this.props.vpcConfig.subnets;
      } else {
        // use all subnets in the VPC
        this.subnets = this.props.vpcConfig.vpc.selectSubnets().subnets;
      }

      const isOnlyIsolatedSubnets = this.subnets.every((subnet) => this.props.vpcConfig?.vpc.isolatedSubnets.includes(subnet));
      const hasS3VpcEndpoint = App.of(this)?.node.findAll().some((resource) => {
        return resource instanceof ec2.CfnVPCEndpoint
          && resource.serviceName === `com.amazonaws.${Stack.of(this).region}.s3`
          && resource.vpcId === this.props.vpcConfig?.vpc.vpcId;
      });
      if (isOnlyIsolatedSubnets && !hasS3VpcEndpoint) {
        throw new Error('VPC configuration must include at least one subnet that is not an isolated subnet or a VPC endpoint to S3');
      };
    }
```

> I think we can add the bucket policy restricting access to the VPC without issues.
I'm sorry. You are right.

> what about adding a condition that enforces it only on the task role?
Nice! I'll implement it.